### PR TITLE
[WIP] borrowck diagnostic migration with eager booted

### PR DESCRIPTION
**Pull request by AndyJado**
on top of #104055

### CODE REVIEW

File: compiler/rustc_borrowck/src/session_diagnostics.rs [hunk 1]
```diff
@@ -240,3 +304,546 @@ pub(crate) struct MoveBorrow<'a> {
...
+    pub kind_new: &'a str,
+    pub noun_old: &'a str,
+    pub kind_old: &'a str,
+    pub is_msg_old_empty: bool,
+    pub msg_old: &'a str,
+    #[primary_span]
+    pub span: Span,
+    #[label]
+    pub old_load_end_span: Option<Span>,
+    #[subdiagnostic(eager)]
```
Comment (AndyJado) [inline compiler/rustc_borrowck/src/session_diagnostics.rs:829]: broke by #105233 @mejrs

Comment (mejrs) [reply compiler/rustc_borrowck/src/session_diagnostics.rs:829]: I agree something like `BorrowOccurLabel` would benefit from `(eager)`. let me see if I can put that back...

As an aside, looking at the `cannot_reborrow_already_borrowed` function, I see that it has only three call locations, and they either invoke this with (immutable, mutable) or (mutable, immutable). You can simplify `ReborrowBorrowedErr` a lot if you split it up into `CannotBorrowMutAlreadyBorrowedImmutErr` and `CannotBorrowImmutAlreadyBorrowedMutErr`.

Comment (AndyJado) [reply compiler/rustc_borrowck/src/session_diagnostics.rs:829]: my flawful brain failed to parse `CannotBorrowMutAlreadyBorrowedImmutErr`..

anyway, I could put `(eager)` back if you don't feel like it, been messing around with macro a lot lately

Comment (mejrs) [reply compiler/rustc_borrowck/src/session_diagnostics.rs:829]: I've opened a PR :)

**Comment by davidtwco** [comment]
I'll look at this after we land #104055

File: compiler/rustc_borrowck/src/borrowck_errors.rs [hunk 2]
```diff
@@ -1,9 +1,20 @@
+// #![deny(rustc::untranslatable_diagnostic)]
+// #![deny(rustc::diagnostic_outside_of_impl)]
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/borrowck_errors.rs:2]: If the code can't compile with these uncommented then we can remove the comments until it can :)

File: compiler/rustc_borrowck/messages.ftl [hunk 3]
```diff
@@ -172,37 +500,59 @@ borrowck_suggest_create_freash_reborrow =
...
+    used here by closure
+
 borrowck_used_impl_require_static =
     the used `impl` has a `'static` requirement
 
+borrowck_used_non_generic_for_generic =
+    used non-generic {$descr ->
+    [lifetime] lifetime
+    [type] type
+    *[constant] constant
```
Comment (davidtwco) [inline compiler/rustc_borrowck/messages.ftl:537]: Ideally we'd split this into three messages rather than pattern matching on a string. We anticipate it would be difficult for translators to know what to match for in their translations. Likewise anywhere else you do this.

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 4]
```diff
@@ -1566,6 +1566,7 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
     /// cannot borrow `a.u` (via `a.u.z.c`) as immutable because it is also borrowed as
     /// mutable (via `a.u.s.b`) [E0502]
     /// ```
+    // FIXME: first 3 out of 4 can contain English word value, note for diagnostic migration
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:1569]:
```suggestion
    // FIXME(#100717): In the return value, the first three strings can contain untranslated text.
```

File: compiler/rustc_borrowck/src/diagnostics/mod.rs [hunk 5]
```diff
@@ -707,21 +708,22 @@ impl<'tcx> BorrowedContentSource<'tcx> {
         }
     }
 
-    pub(super) fn describe_for_immutable_place(&self, tcx: TyCtxt<'_>) -> String {
+    // ready to remove
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mod.rs:711]: Then remove it :)

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 6]
```diff
@@ -24,6 +27,22 @@ pub(crate) enum AccessKind {
     Mutate,
 }
 
+#[derive(Debug)]
+pub(crate) enum PlaceAndReason<'a> {
+    // place, name
+    DeclaredImmute(String, Option<String>),
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:33]: It would be good if we could take the `describe_any_place` logic and put it into a `IntoDiagnosticArg` so that types like this could contain the `Place`s themselves.

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 7]
```diff
@@ -42,20 +61,20 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
 
         let mut err;
         let item_msg;
-        let reason;
+        let diagnosing: PlaceAndReason<'_>;
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:64]:
```suggestion
        let diagnostic: PlaceAndReason<'_>;
```

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 8]
```diff
@@ -173,7 +188,10 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
             } => bug!("Unexpected immutable place."),
         }
 
-        debug!("report_mutability_error: item_msg={:?}, reason={:?}", item_msg, reason);
+        debug!(
+            "report_mutability_error: item_msg and reason is moved to a struct due to diagnosing migration: {:?}",
+            &diagnosing
+        );
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:194]:
```suggestion
        debug!(?diagnosing);
```
nit: if it makes sense to keep this debug output

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 9]
```diff
@@ -183,9 +201,14 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         let mut mut_error = None;
         let mut count = 1;
 
+        let cant_access_here = |sp: Span| match error_access {
+            AccessKind::MutableBorrow => FnMutBumpFn::CannotBorrowMut { sp },
+            AccessKind::Mutate => FnMutBumpFn::CannotAssign { sp },
+        };
         let span = match error_access {
+            // cannot assign only happen in Mutate
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:209]: This comment is a quite confusing, it could easily be describing the behaviour of the compiler, rather than the ability to assign to `diagnosing`.

File: compiler/rustc_borrowck/src/borrowck_errors.rs [hunk 10]
```diff
@@ -29,67 +40,49 @@ impl<'cx, 'tcx> crate::MirBorrowckCtxt<'cx, 'tcx> {
...
-            new_loan_span,
-            E0499,
-            "cannot borrow {}{} as mutable more than once at a time",
-            desc,
-            via(opt_via),
-        );
-        if old_loan_span == new_loan_span {
+    ) -> DiagnosticBuilder<'cx, ErrorGuaranteed> {
+        use crate::session_diagnostics::MutBorrowMulti::*;
+        let via = |msg: &str| msg.is_empty();
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/borrowck_errors.rs:56]: I don't think this closure is necessary.

File: compiler/rustc_borrowck/src/diagnostics/region_name.rs [hunk 11]
```diff
@@ -156,10 +158,11 @@ impl RegionName {
                 );
             }
             RegionNameSource::AnonRegionFromUpvar(span, upvar_name) => {
-                diag.span_label(
-                    *span,
-                    format!("lifetime `{self}` appears in the type of `{upvar_name}`"),
-                );
+                diag.subdiagnostic(RegionNameLabels::LifetimeInTypeOf {
+                    span: *span,
+                    upvar_name: upvar_name.to_ident_string(),
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/region_name.rs:163]: Why do we need `to_ident_string`? `Symbol` should work in the diagnostic, no?

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 12]
```diff
@@ -183,9 +201,14 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         let mut mut_error = None;
         let mut count = 1;
 
+        let cant_access_here = |sp: Span| match error_access {
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:204]: This could use a more descriptive name.

**Review by davidtwco — CHANGES_REQUESTED** [review]
Some comments from a first pass - apologies for the delay in getting to this.

**Comment by Dylan-DPC** [comment]
@AndyJado any updates on this?

**Comment by Dylan-DPC** [comment]
@AndyJado this is no longer blocked, so do you have any updates on it?

**Comment by AndyJado** [comment]
I'm back! I'll get the engine running in a few days.

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs (hunk 9 above)
Comment (AndyJado) [reply compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:209]: It was a note for myself to locate the `cannot_assign()` for debugging a `eager_subdiag` case, I should have removed it.

File: tests/ui/span/borrowck-borrow-overloaded-auto-deref-mut.stderr
Comment (AndyJado) [inline tests/ui/span/borrowck-borrow-overloaded-auto-deref-mut.stderr:1]: @davidtwco I suppose this is a `need eager on Diagonostic (not Subdiag)` case, I've made all my attempts and I can't pass ui test，they are like whack-a-mole

File: compiler/rustc_borrowck/src/diagnostics/mod.rs [hunk 13]
```diff
@@ -174,6 +186,10 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
         }
     }
 
+    pub(super) fn describe_place_typed(&self, place_ref: PlaceRef<'tcx>) -> DescribedPlace {
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mod.rs:189]: I presume this exists because changing `describe_place` to return a DescribedPlace` isn't practical?

File: compiler/rustc_borrowck/src/diagnostics/mod.rs [hunk 14]
```diff
@@ -63,6 +63,18 @@ pub(super) struct DescribePlaceOpt {
...
 pub(super) struct IncludingTupleField(pub(super) bool);
 
+#[derive(Debug)]
+pub(super) struct DescribedPlace(pub(super) Option<String>);
+
+impl IntoDiagnosticArg for DescribedPlace {
+    fn into_diagnostic_arg(self) -> DiagnosticArgValue<'static> {
+        match self.0 {
+            Some(descr) => descr.into_diagnostic_arg(),
+            None => "value".into_diagnostic_arg(),
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mod.rs:73]: Maybe `bug!` here?

File: compiler/rustc_borrowck/src/diagnostics/mod.rs [hunk 15]
```diff
@@ -63,6 +63,18 @@ pub(super) struct DescribePlaceOpt {
 
 pub(super) struct IncludingTupleField(pub(super) bool);
 
+#[derive(Debug)]
+pub(super) struct DescribedPlace(pub(super) Option<String>);
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mod.rs:67]: Why is this necessary if `self.0` already implements `IntoDiagnosticArg`?

File: compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs [hunk 16]
```diff
@@ -120,36 +138,29 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
...
-                                format!(", as it is behind {pointer_type}")
-                            }
-                        }
-                    } else {
-                        item_msg = format!("data in {pointer_type}");
-                        reason = String::new();
-                    }
+                    let name = ptr_source.describe_for_immutable_place(self.infcx.tcx);
+                    diagnostic = PlaceAndReason::BehindPointer(
+                        self.describe_place_typed(access_place.as_ref()),
```
Comment (davidtwco) [inline compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs:159]: Ah, so the "value" case in `DescribedPlace` is to handle the situation of "data" here?

**Review by davidtwco — CHANGES_REQUESTED** [review]
Left some more comments, you'll need to look at where the backticks are being added and make sure that we either only do them in the `IntoDiagnosticArg` implementations or in the Fluent messages.

File: compiler/rustc_borrowck/src/diagnostics/mod.rs (hunk 13 above)
Comment (AndyJado) [reply compiler/rustc_borrowck/src/diagnostics/mod.rs:189]: `describe_place` is used a lot, I'm currently narrowing the scope so it covers only `PlaceAndReason` cases.

**Comment by AndyJado** [comment]
```rust
    fn emit_diagnostic(&mut self, diagnostic: &mut Diagnostic) -> Option<ErrorGuaranteed> {
        dbg!(&diagnostic.args().collect::<Vec<_>>());

```

I tracked function `emit_diagnostic()` and located where backticks are added, and my assumption is that it happens in some macro. 

```
+       [compiler/rustc_errors/src/lib.rs:1326] &diagnostic.args().collect::<Vec<_>>() = [
+           (
+               "place",
+               Str(
+                   "x",
+               ),
+           ),
+       ]
1       error[E0596]: cannot borrow `x` as mutable, as it is a captured variable in a `Fn` closure
2         --> $DIR/issue-21600.rs:14:20
3          |

10         |         |       in this closure
11         |         expects `Fn` instead of `FnMut`
12
-       error[E0596]: cannot borrow `x` as mutable, as it is a captured variable in a `Fn` closure
+       [compiler/rustc_errors/src/lib.rs:1326] &diagnostic.args().collect::<Vec<_>>() = []
+       [compiler/rustc_errors/src/lib.rs:1326] &diagnostic.args().collect::<Vec<_>>() = [
+           (
+               "place",
+               Str(
+                   "`x`",
+               ),
+           ),
+       ]
+       error[E0596]: cannot borrow ``x`` as mutable, as it is a captured variable in a `Fn` closure
```

**Comment by Dylan-DPC** [comment]
@AndyJado any updates on this? thanks

**Comment by AndyJado** [comment]
`cannot_assign` and `cannot_borrow` used to be string interpolation in one big function, after split them into structs E0594 and E0596, backtick repeats on some place description:
```console
- error[E0596]: cannot borrow `r` as mutable
+ error[E0596]: cannot borrow ``r`` as mutable
```
I believe the answer lies in the macro implementation but I don't have the guts diving in. @Dylan-DPC

**Comment by JohnCSimon** [comment]
@AndyJado 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive
